--- a/src/Utils/Utils_ExceptHandlers.hxx
+++ b/src/Utils/Utils_ExceptHandlers.hxx
@@ -38,30 +38,18 @@
 class UTILS_EXPORT Unexpect { //save / retrieve unexpected exceptions treatment
   PVF old;
   public :
-#ifndef WIN32
-    Unexpect( PVF f ) 
+    Unexpect( PVF f )
       { old = std::set_unexpected(f); }
   ~Unexpect() { std::set_unexpected(old); }
-#else
-    Unexpect( PVF f ) 
-      { old = ::set_unexpected(f); }
-  ~Unexpect() { ::set_unexpected(old); }
-#endif
 };
 
 class UTILS_EXPORT Terminate {//save / retrieve terminate function
-  
+
   PVF old;
   public :
-#ifndef WIN32
-    Terminate( PVF f ) 
+    Terminate( PVF f )
       { old = std::set_terminate(f); }
   ~Terminate() { std::set_terminate(old); }
-#else
-    Terminate( PVF f ) 
-      { old = ::set_terminate(f); }
-  ~Terminate() { ::set_terminate(old); }
-#endif
 };
 
 #define UNEXPECT_CATCH(FuncName, ExceptionConstructor) \
