diff --git a/src/NETGENPlugin/src/NETGENPlugin/NETGENPlugin_Mesher.cxx b/src/NETGENPlugin/src/NETGENPlugin/NETGENPlugin_Mesher.cxx
index de79d97..7206875 100644
--- a/src/NETGENPlugin/src/NETGENPlugin/NETGENPlugin_Mesher.cxx
+++ b/src/NETGENPlugin/src/NETGENPlugin/NETGENPlugin_Mesher.cxx
@@ -154,12 +154,11 @@ namespace
     {
       return (( Contains( other.n1 ) || Contains( other.n2 )) && ( this != &other ));
     }
-    static int HashCode(const Link& aLink, int aLimit)
+    size_t operator()(const Link& aLink) const
     {
-      return ::HashCode(aLink.n1 + aLink.n2, aLimit);
+      return static_cast<size_t>(aLink.n1 + aLink.n2);
     }
-
-    static Standard_Boolean IsEqual(const Link& aLink1, const Link& aLink2)
+    bool operator()(const Link& aLink1, const Link& aLink2) const
     {
       return (( aLink1.n1 == aLink2.n1 && aLink1.n2 == aLink2.n2 ) ||
               ( aLink1.n1 == aLink2.n2 && aLink1.n2 == aLink2.n1 ));
