diff --git a/CMakeLists.txt b/CMakeLists.txt
index 65a806d..fa4899d 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -127,7 +127,7 @@ install(FILES ${SMDS_INCLUDES} DESTINATION "include/smesh/SMDS")
 file(GLOB SMESH_SRC ${SMESH_SRC_DIR}/SMESH/*.cxx)
 list(REMOVE_ITEM SMESH_SRC "${SMESH_SRC_DIR}/SMESH/MG_ADAPT.cxx")
 add_library(SMESH ${SMESH_SRC})
-target_link_libraries(SMESH Kernel Geom Controls SMESHDS Driver DriverDAT DriverSTL DriverUNV DriverGMF TKShHealing TKPrim TKG2d TKCDF TKMeshVS ${Boost_LIBRARIES})
+target_link_libraries(SMESH Kernel Geom Controls SMESHDS Driver DriverDAT DriverSTL DriverUNV DriverGMF TKShHealing TKPrim TKG2d TKCDF TKMeshVS TKExpr ${Boost_LIBRARIES})
 target_include_directories(SMESH PUBLIC
   $<BUILD_INTERFACE:${SMESH_SRC_DIR}/SMESH>
   $<INSTALL_INTERFACE:include/smesh/SMESH>
